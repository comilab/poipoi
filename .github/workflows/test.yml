name: test

on:
  push:
    branches:
      - 'master'
      - 'develop'
  pull_request:
    types: [opened, synchronize]
    branches:
      - 'master'
      - 'develop'

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: docker-compose
        run: |
          docker-compose -f docker-compose.ci.yml up -d
      - name: setup-php
        run: |
          docker-compose exec -T php composer install --no-interaction
          docker-compose exec -T php yarn install
          docker-compose exec -T php cp .env.example .env
          docker-compose exec -T php php artisan key:generate
          docker-compose exec -T php touch database/database.sqlite
      - name: setup-client
        run: |
          docker-compose exec -T node yarn install
      - name: test-php
        run: docker-compose exec -T php php artisan test
      - name: test-client
        run: docker-compose exec -T node yarn test
