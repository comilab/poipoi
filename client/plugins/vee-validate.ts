import Vue from 'vue'
import { ValidationObserver, ValidationProvider, extend, localize } from 'vee-validate'
import * as rules from 'vee-validate/dist/rules'
import ja from 'vee-validate/dist/locale/ja.json'

for (const [rule, validation] of Object.entries(rules)) {
  extend(rule, {
    ...validation as any
  })
}

localize({
  ja
})
localize('ja')

Vue.component('ValidationObserver', ValidationObserver)
Vue.component('ValidationProvider', ValidationProvider)
