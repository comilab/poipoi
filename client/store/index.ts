import { defineComponent, provide, inject } from 'nuxt-composition-api'
import { InjectionKey } from '@vue/composition-api'

import sessionStore from '~/store/session'
import settingStore from '~/store/setting'
import toastStore from '~/store/toast'

type Store = {
  session: ReturnType<typeof sessionStore>
  setting: ReturnType<typeof settingStore>
  toast: ReturnType<typeof toastStore>
}

const makeStore = () => {
  return {
    session: sessionStore(),
    setting: settingStore(),
    toast: toastStore()
  }
}

export const key: InjectionKey<Store> = Symbol('Store')

export const store: Store = makeStore()

export const StoreProvider = defineComponent({
  setup () {
    provide(key, store)
  },
  render (h) {
    return h('div', this.$slots.default)
  }
})

export const useStore = () => inject(key)!

export const resetStore = () => {
  Object.assign(store, makeStore())
}

export const actions = {
  async nuxtClientInit () {
    try {
      await Promise.all([
        store.session.load(),
        store.setting.load()
      ])
    } catch (error) {
      //
    }
  }
}
