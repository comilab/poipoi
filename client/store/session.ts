import { reactive, toRefs, computed } from 'nuxt-composition-api'
import axios from 'axios'

import User from '~/models/User'

export default function sessionStore () {
  const state = reactive({
    user: null as User|null
  })

  const isLogin = computed(() => !!state.user)

  async function create (query: any) {
    await axios.get('/sanctum/csrf-cookie', {
      baseURL: process.env.apiUrl
    })
    await axios.post('/sessions', {
      ...query,
      device_name: 'poipoi-web'
    })
  }

  async function load () {
    try {
      const { data } = await axios.get('/sessions')
      setUser(new User(data))
    } catch (error) {
      setUser(null)
    }
  }

  async function destroy () {
    try {
      await axios.delete('/sessions/poipoi-web')
    } catch (error) {
      //
    }
    setUser(null)
  }

  function setUser (user: User|null) {
    state.user = user
  }

  return {
    ...toRefs(state),
    isLogin,
    create,
    load,
    destroy,
    setUser
  }
}
