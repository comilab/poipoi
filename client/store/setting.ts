import { reactive, toRefs } from 'nuxt-composition-api'
import axios from 'axios'

import useApi from '~/composables/use-api'
import Setting from '~/models/Setting'

export default function sessionStore () {
  const api = useApi('/settings', Setting)

  const state = reactive({
    data: null as Setting|null
  })

  async function load () {
    const { data } = await axios.get('/settings')
    setData(new Setting(data))
  }

  async function save (data: any) {
    await api.save(data)
    setData(api.item.value!)
  }

  function setData (data: Setting|null) {
    state.data = data
  }

  return {
    ...toRefs(state),
    load,
    save,
    setData
  }
}
