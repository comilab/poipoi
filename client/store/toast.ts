import { reactive, toRefs } from 'nuxt-composition-api'

export type Toast = {
  type: 'success'|'info'|'warning'|'danger'
  message: string
  timeout?: number
}

export default function toastStore () {
  const state = reactive({
    toasts: [] as Toast[]
  })

  function push ({ type, message, timeout = 5000 }: Toast) {
    const toast = {
      type,
      message,
      timeout
    }
    state.toasts.push(toast)
    if (timeout) {
      setTimeout(() => {
        remove(toast)
      }, timeout)
    }
  }

  function remove (toast: Toast) {
    const index = state.toasts.findIndex(t => t === toast)
    if (index >= 0) {
      state.toasts.splice(index, 1)
    }
  }

  return {
    ...toRefs(state),
    push,
    remove
  }
}
