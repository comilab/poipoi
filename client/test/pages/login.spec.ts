import { fireEvent, render, waitFor, within } from '@testing-library/vue'
import { mocked } from 'ts-jest/utils'
import * as NuxtCompositionApi from 'nuxt-composition-api'
import { Ref } from '@vue/composition-api'
import { validate } from 'vee-validate'
import faker from 'faker'

import Login from '~/pages/login.vue'
import * as store from '~/store'
import { getRules } from '~/test/helper'
import useValidationObserverStub from '~/test/stubs/ValidationObserver'

describe('pages/login', () => {
  let options: any
  let contextMock: any

  beforeEach(() => {
    options = {
      stubs: {
        FontAwesomeIcon: true
      }
    }

    jest.spyOn(store, 'useStore').mockReturnValue(store.store)

    contextMock = {
      redirect: jest.fn()
    }
    jest.spyOn(NuxtCompositionApi, 'useContext').mockReturnValue(contextMock)
  })

  test('renders correctly', () => {
    const { container } = render(Login, options)

    expect(container).toMatchSnapshot()
  })

  describe('form', () => {
    describe('inputs', () => {
      describe('メールアドレス', () => {
        let rules: any

        beforeEach(() => {
          rules = getRules(Login, 'メールアドレス', options)!
        })

        test('必須', async () => {
          const result1 = await validate('', rules)
          expect(result1.valid).toBeFalsy()
          expect(result1.failedRules.required).toBeTruthy()

          const result2 = await validate(faker.internet.email(), rules)
          expect(result2.valid).toBeTruthy()
        })

        test('メールアドレス', async () => {
          const result1 = await validate(faker.random.word(), rules)
          expect(result1.valid).toBeFalsy()
          expect(result1.failedRules.email).toBeTruthy()

          const result2 = await validate(faker.internet.email(), rules)
          expect(result2.valid).toBeTruthy()
        })
      })

      describe('パスワード', () => {
        let rules: any

        beforeEach(() => {
          rules = getRules(Login, 'パスワード', options)!
        })

        test('必須', async () => {
          const result1 = await validate('', rules)
          expect(result1.valid).toBeFalsy()
          expect(result1.failedRules.required).toBeTruthy()

          const result2 = await validate(faker.random.word(), rules)
          expect(result2.valid).toBeTruthy()
        })
      })
    })

    describe('submit', () => {
      let invalid: Ref<boolean>

      beforeEach(() => {
        const stub = useValidationObserverStub()
        options.stubs.ValidationObserver = stub.component
        invalid = stub.invalid

        jest.spyOn(store.store.session, 'create').mockResolvedValue()
        jest.spyOn(store.store.session, 'load').mockResolvedValue()
      })

      test('入力エラーがあれば送信不可', () => {
        invalid.value = true
        const { getByRole } = render(Login, options)

        expect(getByRole('button', { name: 'ログイン' })).toHaveAttribute('disabled')
      })

      test('入力エラーがなければ送信可能', () => {
        const { getByRole } = render(Login, options)

        expect(getByRole('button', { name: 'ログイン' })).not.toHaveAttribute('disabled')
      })

      test('送信中は送信不可', async () => {
        const { getByRole } = render(Login, options)
        await fireEvent.submit(getByRole('form'))

        expect(getByRole('button', { name: 'ログイン' })).toHaveAttribute('disabled')
      })

      test('送信中はindeterminate = true', async () => {
        const { getByRole } = render(Login, options)
        expect(within(getByRole('button', { name: 'ログイン' })).queryByRole('status')).not.toBeInTheDocument()

        await fireEvent.submit(getByRole('form'))

        expect(within(getByRole('button', { name: 'ログイン' })).queryByRole('status')).toBeInTheDocument()
      })
    })

    describe('onSubmit', () => {
      beforeEach(() => {
        jest.spyOn(store.store.session, 'create').mockResolvedValue()
        jest.spyOn(store.store.session, 'load').mockResolvedValue()
      })

      test('store.session.createを実行', async () => {
        const { getByRole } = render(Login, options)
        await fireEvent.submit(getByRole('form'))

        expect(store.store.session.create).toBeCalled()
      })

      test('store.session.loadを実行', async () => {
        const { getByRole } = render(Login, options)
        await fireEvent.submit(getByRole('form'))

        await waitFor(() => {
          expect(store.store.session.load).toBeCalled()
        })
      })

      test('リクエストに成功したらリダイレクト', async () => {
        const { getByRole } = render(Login, options)
        await fireEvent.submit(getByRole('form'))

        await waitFor(() => {
          expect(contextMock.redirect).toBeCalledWith(200, '/')
        })
      })

      test('リクエストに失敗したらエラーを表示', async () => {
        mocked(store.store.session.create).mockRejectedValue(new Error())
        const { getByRole, getAllByText } = render(Login, options)
        await fireEvent.submit(getByRole('form'))

        await waitFor(() => {
          expect(getAllByText('メールアドレスかパスワードが違います')).toHaveLength(2)
        })
      })
    })
  })
})
