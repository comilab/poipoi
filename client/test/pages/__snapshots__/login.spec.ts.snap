// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/login renders correctly 1`] = `
<div>
  <section
    class="rounded bg-white my-8 p-4 max-w-screen-sm mx-auto"
    data-testid="section"
  >
    <h3
      class="text-lg"
      data-testid="title"
    >
      ログイン
    </h3>
    <form
      name="login"
    >
      <div
        class="flex flex-col my-4 untouched pristine"
        data-testid="input-メールアドレス"
      >
        <label
          class="text-sm font-bold py-2"
          data-testid="label"
          for="input-0"
        >
          メールアドレス
        </label>
        <input
          class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 bg-white p-2"
          data-testid="input"
          id="input-0"
          name="メールアドレス"
          type="email"
        />
        <div
          class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
        >
          <div
            data-testid="errors"
          />
          <!---->
        </div>
      </div>
      <div
        class="flex flex-col my-4 untouched pristine"
        data-testid="input-パスワード"
      >
        <label
          class="text-sm font-bold py-2"
          data-testid="label"
          for="input-1"
        >
          パスワード
        </label>
        <input
          class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 bg-white p-2"
          data-testid="input"
          id="input-1"
          name="パスワード"
          type="password"
        />
        <div
          class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
        >
          <div
            data-testid="errors"
          />
          <!---->
        </div>
      </div>
      <div
        class="text-center"
      >
        <button
          class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none bg-blue-600 text-white hover_bg-blue-700 px-3 py-2 rounded"
          data-testid="button"
          type="submit"
        >
          <div>
            <fontawesomeicon-stub
              class="fa-fw mr-2"
              data-testid="icon"
              icon="[object Object]"
            />
          </div>
          <div
            class=""
            data-testid="text"
          >
            ログイン
          </div>
        </button>
      </div>
    </form>
  </section>
</div>
`;
